## Language Processing assignment 3: Word embeddings and society

In this assignment you will have to load vectorial representations of words and calculate their cosine similarity, a common distance metric to evaluate semantic similarity.

On grading: There are five exercises in this assignment. You must have at least three correct exercises (and among the incorrect ones, there should be some proper attempt to solve the missing exercises). What we mean is that if you do three perfect exercises but the remaining two exercises are blank, the assignment will not be considered passed.

In [1]:
import numpy as np
import math

 #### Exercise 1:
 
In order to play with word embeddings, we need a way of storing them in our program. We need a data structure to represent all the word embeddings.

The goal of this exercise is to open the file where the embeddings are saved and to put them in a variable that you can use afterwards.

You can represent the data in the way that you think it fits best. The result can go from a really simple approach until a complex but useful class.

Given a word, such as `"house"`, this data structure should return the embeddings related to that word.

In the saved file, we will have a set of words, and each word will be represented as a sequence of floating point numbers, such as:

`house --> 0.001 0.002 0.005 0.001 0.0012312 0.004 ...`

`cow --> 0.2 0.01 0.00031 0.01 0.9 0.00031 0.0015 0.002 ...`

The number of floating point numbers will always be the same.

---
 
If we want to play with word embeddings, we have to get them from somewhere. Pick English embeddings from [Absalon](https://absalon.ku.dk/files/7371057/download?download_frd=1) or the embeddings that you want from this [website](https://fasttext.cc/docs/en/crawl-vectors.html). I recommend you downloading the file from Absalon, as it contains only 50,000 words and it is easier to load (well, faster).

If you get the embeddings from the Github page, you should download the embeddings in **text** format. These embeddings have been trained with raw text from Wikipedia. This may take some space in your computer, depending on the language you choose.

Once you downloaded the embeddings, it's time to start programming! The files follow a specific format.

##### FILE FORMAT:

The first line in the file contains two numbers separated by a single space. The first number indicates the number of words in the file (`N_WORDS`) and the second number specifies the number of dimensions (`N_DIMENSIONS`) that are used to represent each of those words.

After the first line, each line will contain one word at the beginning. Following the word, and separated by spaces, there will be `N_DIMENSIONS` numbers, which represent each word in the space.

The words are sorted by their frequency in the wikipedia corpus, then the first words in the file will be the most frequent ones. Here you can see how the English embeddings file starts:

`9999 300`

`, 0.1250 -0.1079 0.0245 -0.2529 0.1057 -0.0184 0.1177 ...`

`the -0.0517 0.0740 -0.0131 0.0447 -0.0343 0.0212 0.0069 ...`

`. 0.0342 -0.0801 0.1162 -0.3968 -0.0147 -0.0533 0.0606 ...`

`and 0.0082 -0.0899 0.0265 -0.0086 -0.0609 0.0068 0.0652 ...`

`...`

##### What you have to do:

Write a program to read the file and store the words and their embeddings in the data structure that you think it is the best. It might be very simple, or it might be a more complex one.

##### Important note: You are not allowed to use a package like gensim to open the file

In [2]:
#1.- Define object to save words and their embeddings
#2.- Write code for reading the file and save it in the defined object
#YOUR CODE HERE
words = dict()
with open("wiki.en.vec.short50K") as f:
    f.readline() # skip first line
    for line in f:
        line = line.split()
        vector = np.array(line[1:], dtype=float) 
        if len(vector) == 300: # only keep vectors of size 300
            word = line[0] 
            words[word] = vector # add word and vector to words dictionary
print(len(words)) 

49997


#### Exercise 2:

A common distance metric used to measure the similarity between two words is the cosine similarity, which measures the cosine of the angle between the two vectors that represent each of the words.

This similarity value is calculated by using this formula:

$$\text{similarity} = \cos(\theta) = {\mathbf{A} \cdot \mathbf{B} \over \|\mathbf{A}\|_2 \|\mathbf{B}\|_2} $$

<!--= \frac{ \sum\limits_{i=1}^{n}{A_i  B_i} }{ \sqrt{\sum\limits_{i=1}^{n}{A_i^2}}  \sqrt{\sum\limits_{i=1}^{n}{B_i^2}} }-->

Don't be scared. The first part of the formula, $\mathbf{A} \cdot \mathbf{B}$ is the dot product between vectors $\mathbf{A}$ and $\mathbf{B}$. And you know how to do that in Python.

$\mathbf{A} \cdot \mathbf{B} = \sum\limits_{i=1}^{n}{A_i  B_i}$

In the lower part, $\|\mathbf{A}\|_2 \|\mathbf{B}\|_2$, you have to calculate the Euclidean norm of each vector ($\mathbf{A}$ and $\mathbf{B}$) and multiply their results. The Euclidean norm is calculated using this formula:

$\|\mathbf{A}\|_2 = \sqrt{\sum\limits_{i=1}^{n}{A_i^2}}$

The formula inside the square root is the same as the one from the dot product. Then it can be rewritten like this:

$\|\mathbf{A}\|_2 = \sqrt{\sum\limits_{i=1}^{n}{A_i^2}} = \sqrt{\mathbf{A} \cdot \mathbf{A}} $

You should program the cosine similarity function by using numpy. You cannot use previously programmed cosine similarity functions, you must write your own function. This program must get two numpy arrays and it should return a number.

The resulting number of this formula should be interpreted as a number that specifies the similarity between two words. The higher the number, the similarity between those two words will be higher.

In [3]:
def similarity(A, B):
    #YOUR CODE HERE
    return np.dot(A,B)/(np.sqrt(np.dot(A,A))*np.sqrt(np.dot(B,B)))

similarity(words["and"],words["but"])

0.5719145601185748

#### Exercise 3:

In the third exercise you have to squeeze your brain a bit more. Now, you have loaded the whole embedding file, and you also have a distance metric to measure the similarity between words. Let's do more complex things, then.

Given a word, you have to find the 30 most similar words. Then, given one word you should get the distance to all the words in the embeddings file, and pick the nearest ones.

In order to make this task easier, I attach a simple implementation of an ordered list.

In [4]:
#This function should return the embeddings of a word according to your class
def get (LIST, index):
    return LIST[index]

def get_value(el):
    return el[1]



class OrderedListTuple:
    
    def __init__(self, max_size):
        self.content = []
        self.max_size = max_size
        
    def find_pos (self, element):
        index = 0
        while (index <= len(self.content)-1) and get_value(get(self.content, index)) > get_value(element):
            index += 1
        return index

    def insert_element (self, element):
        pos = self.find_pos (element)
        self.content.insert (pos, element)
        if len(self.content) > self.max_size:
            self.content.pop()

This implementation is very simple. When we initialize the list, we set the number of elements that it will have at most. Then, when we add elements to the list, it will add the element in the correct position. But, if the number of elements is higher than the ones that we can keep, the object will remove the last element. Let's see how it works with an example:

In [5]:
L = OrderedListTuple(4)
print (L.content)

L.insert_element(("house", 14))
print (L.content)
L.insert_element(("home", 6))
print (L.content)
L.insert_element(("brown", 3))
print (L.content)
L.insert_element(("elbow", 4))
print (L.content)
L.insert_element(("high", 1))
print (L.content)
L.insert_element(("the", 9))
print (L.content)
L.insert_element(("and", 43))
print (L.content)
L.insert_element(("kitty", 44))
print (L.content)

[]
[('house', 14)]
[('house', 14), ('home', 6)]
[('house', 14), ('home', 6), ('brown', 3)]
[('house', 14), ('home', 6), ('elbow', 4), ('brown', 3)]
[('house', 14), ('home', 6), ('elbow', 4), ('brown', 3)]
[('house', 14), ('the', 9), ('home', 6), ('elbow', 4)]
[('and', 43), ('house', 14), ('the', 9), ('home', 6)]
[('kitty', 44), ('and', 43), ('house', 14), ('the', 9)]


##### Hint: Why don't you create a similarity function that gets two words, and it returns a tuple? For each word in the dictionary, you can calculate the similarity to an input word, and save this in a tuple. Then, using the previous data structure, you can save only the N-best words.

With this data structure, you should be able to get the most similar words to one word.

In [6]:
#YOUR CODE HERE

def similarity_tuple(word1,word2,dictionary):
    '''
    Calculate similarity between two words based on their vectors.
    
    Parameters:
    - word1 (str): First word.
    - word2 (str): Second word.
    - dictionary (dict): A dictionary that maps words to vectors.
    
    Returns:
    tuple: A tuple containing the second word and the similarity score to the first word.
    '''
    vector1 = dictionary[word1]
    vector2 = dictionary[word2]
    return (word2,similarity(vector1,vector2))

def most_similar_words(input_word,dictionary,max_size):
    '''
    Find the most similar words to a given input word based on similarity between vectors.
    
    Parameters:
    - input_word (str): The input word.
    - dictionary (dict): A dictionary that maps words to vectors.
    - max_size (int): The number of similar words to retrieve.
    '''
    L = OrderedListTuple(max_size)
    for word,vector in dictionary.items():
        similarity = similarity_tuple(input_word,word,dictionary) # get similarity tuple
        L.insert_element(similarity) # insert similarity tuple to L
    return L.content

most_similar_words("and",words,30)

[('and', 1.0),
 ('both', 0.7088399055288035),
 ('well', 0.6507605201793852),
 ('including', 0.6455951558139696),
 ('while', 0.6344329391819657),
 ('lastly', 0.6124539704969483),
 ('additionally', 0.6103882770258247),
 ('the', 0.60341005556975),
 ('of', 0.600915614446954),
 ('which', 0.599827363344629),
 (',', 0.5975515358044979),
 ('along', 0.5920798889560756),
 ('notably', 0.5828809034774284),
 ('also', 0.5823754688131197),
 ('addition', 0.5804041956874957),
 ('other', 0.5773918231698155),
 ('but', 0.5719145601185748),
 ('likewise', 0.5702313752171171),
 ('however', 0.5635267429063764),
 ('with', 0.5604893133079214),
 ('especially', 0.5600702339008268),
 ('respectively', 0.5507238976794473),
 ('these', 0.5502342758557172),
 ('consequently', 0.546646170769302),
 ('besides', 0.545569494198823),
 ('many', 0.5448865850382615),
 ('similarly', 0.543413204922898),
 ('several', 0.5432807476331165),
 ('although', 0.5431728976115298),
 ('furthermore', 0.5383040086371182)]

#### Exercise 4:

The last exercise is really cool. One of the properties that researchers found in word embeddings was that we could perform algebraic operations over the vectors in order to get specific words.

For example, if we perform an operation like this one:

$$DICTIONARY['berlin'] - DICTIONARY['germany'] + DICTIONARY['france']$$

This results in a vector. If we find the 20 closest words to that vector, we should be able to see that the word `Germany` will be near. Another nice operation was:

$$DICTIONARY['queen'] - DICTIONARY['woman'] + DICTIONARY['man']$$

Perform this operations with the words you want, and check if it works.

In [7]:
#YOUR CODE HERE

# modify functions from above so they take vectors as input and not words
def similarity_tuple_vec(vec1,word2,dictionary):
    vec2 = dictionary[word2]
    return (word2,similarity(vec1,vec2))

def most_similar_words_vec(input_vector,dictionary,max_size):
    L = OrderedListTuple(max_size)
    for word,vector in dictionary.items():
        similarity = similarity_tuple_vec(input_vector,word,dictionary)
        L.insert_element(similarity)
    return L.content


result = words['berlin'] - words['germany'] + words['france'] #berlin - germany + france
most_similar_words_vec(result,words,20)

[('paris', 0.7745821281649259),
 ('berlin', 0.6806360501069306),
 ('france', 0.6234364339620112),
 ('marseille', 0.6068132959916012),
 ('toulouse', 0.6053895238433366),
 ('montpellier', 0.5917006165578014),
 ('rouen', 0.5914574790560799),
 ('avignon', 0.5890789796674709),
 ('rennes', 0.5850185342314214),
 ('ferrand', 0.5844373101108589),
 ('brussels', 0.5783981772734191),
 ('nantes', 0.5750482980258821),
 ('bordeaux', 0.5732881366656535),
 ('marseilles', 0.5714683689465664),
 ('boulogne', 0.5677972274128383),
 ('neuilly', 0.5665287787178916),
 ('grenoble', 0.5660398964645867),
 ('lille', 0.5642594745563221),
 ('provence', 0.5618433958267668),
 ('poitiers', 0.558248266369883)]

In [8]:
#YOUR CODE HERE
#queen - woman + man
result = words['queen'] - words['woman'] + words['man']
most_similar_words_vec(result,words,20)

[('queen', 0.7898672082967786),
 ('king', 0.6429407513059967),
 ('majesty', 0.5372264198483585),
 ('monarch', 0.5126070187949001),
 ('crown', 0.469760687497465),
 ('queens', 0.4569217657524951),
 ('whitehall', 0.4508493631090601),
 ('reign', 0.4498360282763127),
 ('coronation', 0.44729696911956957),
 ('royal', 0.43049023386023016),
 ('regent', 0.4278365170631544),
 ('jubilee', 0.42361938568326163),
 ('kings', 0.42048129263683504),
 ('prince', 0.4184276772487855),
 ('connaught', 0.4147187331471937),
 ('consort', 0.4101795096405898),
 ('princess', 0.4090483186856231),
 ('throne', 0.40795276082669685),
 ('pretender', 0.40745142030398634),
 ('elizabeth', 0.4070293809206282)]

#### Exercise 5:

In recent years, many researchers have shown that word embeddings obtained from large corpora reproduce biases that happen in society.

In this exercise, we would like to ask you to try to show some examples in the loaded word embedding file that show some sort of bias. This bias can be either of these, or any other bias you are interested:

 * Gender
 * Origin
 * Sexual preference
 * Socioeconomic class
 * Academic background
 
These examples could be based on distances between words, but any other creative methodology that you could think of will be well considered as well.

For instance, what is the distance between "maid" and "man", and "maid" and "woman"?

You should provide examples but also your interpretation of these results.

If you want to get some inspiration, you may want to check some recent articles about the topic:

  * Bender, Emily M., and Batya Friedman. "Data statements for natural language processing: Toward mitigating system bias and enabling better science." Transactions of the Association for Computational Linguistics 6 (2018): 587-604. https://aclanthology.org/Q18-1041/
  * Hovy, Dirk, and Shrimai Prabhumoye. "Five sources of bias in natural language processing." Language and Linguistics Compass 15.8 (2021): e12432. https://compass.onlinelibrary.wiley.com/doi/10.1111/lnc3.12432

In [9]:
#YOUR CODE HERE

# gender bias (based on distance between words)

print("nurse, man: ", similarity(words["nurse"],words["man"]))
print("nurse, woman: ", similarity(words["nurse"],words["woman"])) 
print()
print("boss, man: ", similarity(words["boss"],words["man"]))
print("boss, woman: ", similarity(words["boss"],words["woman"]))
print()

# ethnicity/origin bias (based on distance between words)

print("scientist, American: ",similarity(words["scientist"],words["american"]))
print("scientist, African: ", similarity(words["scientist"],words["african"]))
print()
print("horrible, Germany: ", similarity(words["horrible"],words["germany"])) # due to historical events, one country has higher sim. with negative words
print("horrible, Netherlands: ", similarity(words["horrible"],words["netherlands"]))
print()

print("scientist, American: ",similarity(words["scientist"],words["american"]))
print("scientist, African: ", similarity(words["scientist"],words["african"]))
print()

nurse, man:  0.21655707042157213
nurse, woman:  0.48998861722812437

boss, man:  0.41120083668655244
boss, woman:  0.2984269239418242

scientist, American:  0.2633526803036764
scientist, African:  0.1314031374973034

horrible, Germany:  0.15864924391812563
horrible, Netherlands:  0.1064987566348144

scientist, American:  0.2633526803036764
scientist, African:  0.1314031374973034



YOUR INTERPRETATION HERE (press ENTER to write)

As indicated by differences in word similarities, the loaded word embedding file shows some forms of gender and ethnicity/origin bias.

Regarding gender bias, the embedding of "nurse" is more similar to the embedding of "woman" than to the embedding of "man". Another example concerns the higher similarity of "boss" and "man" than the similarity of "boss" and "woman". This emphasizes that the dataset reflects stereotypical and historical associations of professions and gender.

Moreover, instances of origin or ethnicity bias could be detected: the embedding for "scientist" is more similar to the embedding of "American" than to the embedding of "African". This could be explained by an overrepresentation of certain cultures and regions in the training data, such as a higher amount of entries describing the discoveries of American scientists. Another example might be the higher similarity of the word pair "horrible"-"Germany" compared to the similarity of "horrible"-"Netherlands". This could be due to negative historical events described in several wikipedia entries. Even though the data was trained on a Wikipedia Corpus and thus consists of texts that aim to describe topics in a more neutral way, biases can still emerge due to, among other factors, the biases of authors of the articles. 

#### Exercise 6

In this exercise we are going to see how to calculate word counts normalized by document frequency (TF-IDF).

To this end, we will calculate the word frequencies and the IDF normalized counts using a Python package called TFIDFvectorizer from Scikit-Learn.

We will work with the Gutenberg corpus from NLTK.

In [10]:
from nltk.corpus import gutenberg
import nltk

import numpy as np

from sklearn.feature_extraction.text import TfidfVectorizer

In [11]:
fileids = gutenberg.fileids()
fileids

['austen-emma.txt',
 'austen-persuasion.txt',
 'austen-sense.txt',
 'bible-kjv.txt',
 'blake-poems.txt',
 'bryant-stories.txt',
 'burgess-busterbrown.txt',
 'carroll-alice.txt',
 'chesterton-ball.txt',
 'chesterton-brown.txt',
 'chesterton-thursday.txt',
 'edgeworth-parents.txt',
 'melville-moby_dick.txt',
 'milton-paradise.txt',
 'shakespeare-caesar.txt',
 'shakespeare-hamlet.txt',
 'shakespeare-macbeth.txt',
 'whitman-leaves.txt']

In [12]:
corpus = []
for fileid in fileids:
    corpus.append(nltk.corpus.gutenberg.raw(fileid))

In [13]:
#Step 1:
vec_tfidf = TfidfVectorizer()
result_tfidf = vec_tfidf.fit_transform(corpus).toarray()

#Step 2:
id2word = {vec_tfidf.vocabulary_[key]:key for key in vec_tfidf.vocabulary_.keys()}

#Step 3:
sorted_ids_blake = np.argsort(result_tfidf[4]).reshape(-1)

#Step 4:
for id in sorted_ids_blake[-10:]:
    print (id2word[id],result_tfidf[4][id])

he 0.08135534295322783
his 0.08280811693453548
thee 0.08905556490592986
with 0.09588308276630424
my 0.1205802404485341
to 0.16125791192514802
in 0.20484113136437723
of 0.21210500127091542
and 0.5055653454950587
the 0.6377677777940539


##### Exercise 6.1:
Explain using your own words and in one single sentence (per step) what happens in each step.

###### Step 1: A TfidfVectorizer is applied to the corpus (list of raw texts from the Gutenberg corpus), thereby creating a TF-IDF matrix (rows=documents, columns=words).

###### Step 2: A dictionary mapping feature indices to terms is created. 

###### Step 3: The indices of the words of  the 5th document ('blake-poems.txt') are sorted in ascending order by TF-IDF score and stored in a 1D array.

###### Step 4: The 10 words with the highest TF-IDF score in the 5th document are printed along with their TF-IDF score.

#### Exercise 6.2:
Can you check what are the top-10 most relevant words based on their inverse document frequency? I am asking for the 10 words with the highest inverse document frequency.

If you do not know how to get the IDFs of the words, you may want to take a look at the documentation of the TFIDF vectorizer.

In [14]:
#YOUR CODE HERE
idfs = vec_tfidf.idf_ # get IDFs of words

sorted_ids_idfs = np.argsort(idfs) # sort ids of words by IDF score

for i in sorted_ids_idfs[-10:]:
    print (id2word[i],idfs[i]) # print the 10 words with highest IDF scores along with their IDF score

hummingly 3.2512917986064953
hummer 3.2512917986064953
hummed 3.2512917986064953
humiliations 3.2512917986064953
humiliates 3.2512917986064953
humh 3.2512917986064953
humboldt 3.2512917986064953
humbling 3.2512917986064953
humblest 3.2512917986064953
zuzims 3.2512917986064953


In [15]:
idf_score = math.log((1 + 18) / (1 + 1)) + 1 # check value for hapax legomena
# formula from: https://scikit-learn.org/stable/modules/generated/sklearn.feature_extraction.text.TfidfTransformer.html, retrieved 9th of Dec 2023
print(idf_score) 

3.2512917986064953


#### Exercise 6.3:

How do the inverse document frequencies look like in this corpus? Do they seem relevant? Please state that in 1-2 sentences.

#### YOUR RESPONSE HERE
The 10 words with highest inverse document frequency all have a value of 3.25, suggesting that they are hapax legomena, so they only appear in one document. If our search query would be "Humboldt" then we would like to retrieve the only document mentioning "Humboldt", so we want to assign a higher weight to that term since it is highly informative. 